File: 096_Minimum Genetic Mutation.py
"""
Minimum Genetic Mutation
Find minimum mutations from start to end using BFS over valid bank strings.
"""
from collections import deque
from typing import List

def min_mutation(start: str, end: str, bank: List[str]) -> int:
    bank_set = set(bank)
    if end not in bank_set: return -1
    chars = 'ACGT'
    q = deque([(start,0)])
    vis = {start}
    while q:
        s,d = q.popleft()
        if s == end: return d
        for i in range(len(s)):
            for c in chars:
                if c == s[i]: continue
                t = s[:i] + c + s[i+1:]
                if t in bank_set and t not in vis:
                    vis.add(t); q.append((t,d+1))
    return -1

if __name__ == '__main__':
    print('Min mutation:', min_mutation('AACCGGTT','AACCGGTA',['AACCGGTA']))

# Time Complexity: O(N * L * 4)
# Space Complexity: O(N)

File: 096_Minimum Genetic Mutation.cpp
// Minimum Genetic Mutation
// Minimum mutations to change start to end using bank; BFS on words graph.
#include <bits/stdc++.h>
using namespace std;

int minMutation(string start, string end, vector<string>& bank){
    unordered_set<string> bankSet(bank.begin(), bank.end());
    if(!bankSet.count(end)) return -1;
    string chars = "ACGT";
    queue<pair<string,int>> q; q.push({start,0}); unordered_set<string> vis{start};
    while(!q.empty()){
        auto [s,d]=q.front(); q.pop();
        if(s==end) return d;
        string t=s;
        for(int i=0;i<s.size();++i){ char old=t[i];
            for(char c: chars){ if(c==old) continue; t[i]=c; if(bankSet.count(t) && !vis.count(t)){ vis.insert(t); q.push({t,d+1}); }}
            t[i]=old;
        }
    }
    return -1;
}

int main(){
    string start="AACCGGTT", end="AACCGGTA"; vector<string> bank={"AACCGGTA"};
    cout<<"Min mutation: "<<minMutation(start,end,bank)<<"\n";
    return 0;
}

// Time Complexity: O(N * L * 4) where N bank size, L word length
// Space Complexity: O(N)
